### PR TITLE
OGR: regenerate shapefile qix on end editing

### LINKED ISSUE #30528: Write a .qix spatial index when exporting vector layer as Shapefile
**Feature description.**
GDAL supports the creation of a .qix spatial index for Shapefiles. As spatial indexes are very useful and beneficial to processing speed in many (if not most) cases of doing any GIS work, a vector layer without a spatial index is less useful than one with. The cost of creating an index is negligible in the vast majority of cases. QGIS automatically utilises such sidecar spatial index on Shapefiles.

QGIS' "Save Vector Layer as.." dialog should expose this option and default to YES (please!).

This small change would probably greatly benefit many users and lead to QGIS being faster in random comparisons *magically* when the resulting files are used for further analyses.

### LINKED ISSUE #42567: Automatically update Spatial Index if layer is edited
**Feature description.**
When a layer that already has a spatial index created is edited the spatial index should automatically be recreated.

**Additional context**
Currently, there is a non-obvious issue when the spatial index no longer matches (ie after the layer has been edited) the layer where points will disappear/reappear depending on zoom level. New points that have been copied to the layer will also not display. Currently there are only two fixes available. 1) Manually remove the spatial index file (an option to delete the index in the properties prompt would also be useful here) or 2) recreate the index after every modification. This can be quite tedious during an extended period layer creation  where there is extensive modification and/or usage of spatial processing (which benefits enormously from the spatial index).

### PR DESCRIPTION
**Pull request by elpaso**
GDAL deletes the qix when saving, if the qix was there before editing started we generate a new one after editing ended.

Maybe fix #42567

### CODE REVIEW

File: src/core/providers/ogr/qgsogrprovider.cpp [hunk 1]
```diff
@@ -35,7 +35,6 @@ email                : sherman at mrcc.com
 #include "qgsembeddedsymbolrenderer.h"
 #include "qgsprovidersublayerdetails.h"
 #include "qgsvectorlayer.h"
-#include "qgsproviderregistry.h"
```
Comment (agiudiceandrea) [inline src/core/providers/ogr/qgsogrprovider.cpp:38]: Was it removed by mistake?

**Comment by nyalldawson** [comment]
> Maybe fix https://github.com/qgis/QGIS/issues/30528

I don't think it does -- this change only means that a spatial index isn't lost upon editing, but it doesn't mean we automatically create one for layers without them.

File: src/core/providers/ogr/qgsogrprovider.cpp (hunk 1 above)
Comment (elpaso) [reply src/core/providers/ogr/qgsogrprovider.cpp:38]: yes and no: I trusted my editor (which uses clang-analyzer) that said it wasn't used, but I see my editor was wrong. I'll put it back.

**Comment by elpaso** [comment]
> > Maybe fix #30528
> 
> I don't think it does -- this change only means that a spatial index isn't lost upon editing, but it doesn't mean we automatically create one for layers without them.

Yeah, sorry, wrong link: https://github.com/qgis/QGIS/issues/42567#issuecomment-879185185

**Review by rouault — APPROVED** [review]
LGTM

**Comment by elpaso** [comment]
@rouault thank you for the review!

**Pull request merged**